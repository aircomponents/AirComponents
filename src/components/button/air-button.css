/* 固定宽度样式 */
.button {
  @apply relative inline-flex items-center justify-center font-medium transition-all duration-200 ease-in-out focus:outline-none;
  border-radius: 0.5%;
  padding: 0.5rem 1rem;
  width: 200px; /* 固定按钮宽度 */
  text-align: center;
  overflow: hidden; /* 超出部分隐藏 */
  text-overflow: ellipsis; /* 超出宽度显示省略号 */
  white-space: nowrap; /* 不换行 */
}

.native-button {
  border-radius: 0.375rem; /* 默认圆角 */
  transition: all 0.2s ease-in-out; /* 增加过渡效果 */
}

/* 如果希望文字换行显示 */
.button.multiline {
  white-space: normal; /* 允许换行 */
  overflow: visible; /* 显示全部内容 */
  text-overflow: unset; /* 不显示省略号 */
}

/* variant 样式 */
.variant-solid {
  @apply text-white bg-blue-500 border border-blue-500;
}

.variant-outline {
  @apply text-blue-500 bg-transparent border border-blue-500;
}

.variant-text {
  @apply text-blue-500 bg-transparent border-none;
}

.variant-neo {
  @apply text-white bg-gradient-to-r from-blue-500 to-teal-500 border-none;
}

.variant-default {
  @apply text-gray-700 bg-gray-100 border border-gray-300;
}

/* 大小样式 */
.size-small {
  @apply px-3 py-1.5 text-sm;
  width: 120px; /* 固定小按钮宽度 */
}

.size-medium {
  @apply px-4 py-2 text-base;
  width: 200px; /* 固定中等按钮宽度 */
}

.size-large {
  @apply px-6 py-3 text-lg;
  width: 300px; /* 固定大按钮宽度 */
}


/* color 样式分类 */
.color-default {
  @apply text-gray-700 bg-gray-100 border border-gray-300;
}

.color-primary {
  @apply text-white bg-blue-500 border border-blue-500;
}

.color-success {
  @apply text-white bg-green-500 border border-green-500;
}

.color-info {
  @apply text-white bg-teal-500 border border-teal-500;
}

.color-warning {
  @apply text-black bg-yellow-400 border border-yellow-500;
}

.color-danger {
  @apply text-white bg-red-500 border border-red-500;
}

.color-ghost {
  @apply text-gray-700 bg-transparent border border-gray-300;
}

/* 动态样式 */
.button:hover:not(:disabled) {
  @apply opacity-90;
}

.button:active:not(:disabled) {
  @apply scale-95;
}

.button:disabled {
  @apply bg-gray-300 text-gray-500 cursor-not-allowed;
}

.button.loading {
  @apply cursor-wait;
}

/* 图标和文本样式 */
.air-button__icon,
.air-button__suffix-icon {
  @apply flex-shrink-0;
}

.air-button__text {
  @apply flex-1 text-center;
}

/* 加载状态动画 */
.air-button__loading-icon {
  @apply animate-spin;
}

/* 确保加载图标、文本和后缀图标适配 */
.air-button__icon,
.air-button__loading-icon,
.air-button__suffix-icon {
  @apply inline-block;
}
