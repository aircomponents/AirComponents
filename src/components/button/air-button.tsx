import { Component, h, Host, Prop, State } from "@stencil/core";

@Component({
  tag: "air-button",
  styleUrl: "air-button.css",
  shadow: true,
})
export class AirButton {
  @Prop() size: "small" | "medium" | "large" = "medium";
  @Prop() variant:
    | "solid"
    | "outline"
    | "text"
    | "outline"
    | "neo"
    | "default" = "solid";
  @Prop() color:
    | "default"
    | "primary"
    | "success"
    | "info"
    | "warning"
    | "danger"
    | "ghost" = "default";
  @Prop() icon: string = "";
  @Prop() suffixIcon: string = "";
  @Prop() disabled: boolean = false;
  @Prop() loading: boolean = false; // 显示加载状态
  @Prop() selected: boolean = false; // 选中状态

  @State() hasSlotContent: boolean = false; // 用来检查按钮插槽内容是否为空

  private nativeElement!: HTMLButtonElement;

  // 检查插槽内容
  private computeSlotHasContent() {
    const slot = this.nativeElement.shadowRoot?.querySelector("slot");
    if (slot) {
      const assignedNodes = slot.assignedNodes();
      this.hasSlotContent =
        assignedNodes.length > 0 || slot.textContent?.trim().length > 0;
    }
  }

  render() {
    const ariaLabel = this.hasSlotContent ? null : "Button"; // 当插槽内容存在时，aria-label 设置为 null

    return (
      <Host>
        <button
          class={{
            "native-button": true,
            [`size-${this.size}`]: true,
            [`variant-${this.variant}`]: true,
            [`color-${this.color}`]: true,
            "loading": this.loading, // 添加加载状态样式
            "disabled": this.disabled, // 添加禁用状态样式
            "selected": this.selected, // 添加选中状态样式
          }}
          aria-busy={this.loading ? "true" : null} // 如果是加载状态，设置 aria-busy
          aria-label={ariaLabel} // 如果插槽内容为空，则使用按钮的默认描述
          title={ariaLabel}
          ref={(elm: HTMLButtonElement) => (this.nativeElement = elm)}
          disabled={this.disabled} // 添加 disabled 属性
        >
          {this.loading ? (
            <span class="air-button__loading-icon">🔄</span> // 如果处于加载状态，显示加载图标
          ) : (
            this.icon && <span class="air-button__icon">{this.icon}</span> // 否则显示按钮图标
          )}
          <slot onSlotchange={() => this.computeSlotHasContent()} />
          {this.suffixIcon && !this.loading && (
            <span class="air-button__suffix-icon">{this.suffixIcon}</span> // 如果有后缀图标并且没有加载状态，则显示后缀图标
          )}
        </button>
      </Host>
    );
  }

  // 组件加载完成后计算插槽内容
  componentDidLoad() {
    this.computeSlotHasContent();
    // 等待渲染完成，避免获取不到 shadowRoot 或 slot 内容
    setTimeout(() => this.computeSlotHasContent(), 0);
  }
}
