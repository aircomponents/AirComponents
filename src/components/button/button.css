/* 核心按钮样式 */
.native-button {
  @apply relative inline-flex items-center justify-center font-medium transition-all duration-200 ease-in-out focus:outline-none;
  @apply px-4 py-2 text-center overflow-hidden gap-2;
  min-height: 2.5rem; /* 最小高度 */
  min-width: 6rem; /* 最小宽度 */
  border-radius: 0.375rem; /* 默认圆角 */
}

/* 按钮尺寸 */
.native-button.size-small {
  @apply px-3 py-1.5 text-sm;
  width: 120px; /* 固定小尺寸宽度 */
}

.native-button.size-medium {
  @apply px-4 py-2 text-base;
  width: 200px; /* 固定中等尺寸宽度 */
}

.native-button.size-large {
  @apply px-6 py-3 text-lg;
  width: 300px; /* 固定大尺寸宽度 */
}

/* 按钮颜色 */
.native-button.color-default {
  @apply text-gray-700 bg-gray-100 border-gray-300;
}

.native-button.color-primary {
  @apply text-white bg-blue-500 border-blue-500;
}

.native-button.color-success {
  @apply text-white bg-green-500 border-green-500;
}

.native-button.color-info {
  @apply text-white bg-teal-500 border-teal-500;
}

.native-button.color-warning {
  @apply text-black bg-yellow-500 border-yellow-500;
}

.native-button.color-danger {
  @apply text-white bg-red-600 border-red-500;
}

.native-button.color-ghost {
  @apply text-gray-700 bg-transparent border-gray-500;
}

/* 按钮变体 */
.native-button.variant-solid {
  @apply border border-solid border-gray-300;
}

.native-button.variant-outline {
  @apply border border-gray-500;
}

.native-button.variant-text {
  @apply bg-transparent border-none;
}

.native-button.variant-neo {
  @apply bg-gradient-to-r from-blue-500 to-teal-500;
}

.native-button.variant-default {
  @apply border;
}

/* 图标和文本样式 */
.native-button__icon,
.native-button__suffix-icon {
  @apply flex-shrink-0; /* 防止图标缩放 */
}

/* 图标适配 */
.native-button__icon,
.native-button__loading-icon,
.native-button__suffix-icon {
  @apply inline-block; /* 图标与文本对齐 */
}

.native-button__text {
  @apply flex-1 text-center min-w-0; /* 文本区域自适应压缩 */
}

/* 加载状态的图标 */
.native-button__loading-icon {
  @apply w-6 h-6 border-4 border-t-transparent border-solid rounded-full animate-spin; /* 加载圆环 */
  border-color: currentColor transparent transparent transparent;
  animation-duration: 1s; /* 动画时长 */
}

.native-button.selected {
  @apply border-2 border-dashed border-blue-500 shadow-md; /* 增加更醒目的蓝色虚线边框 */
  transition: border-color 0.2s ease-in-out, box-shadow 0.2s ease-in-out; /* 边框颜色和阴影的过渡效果 */
}

/* 按钮状态样式 */
.native-button:hover:not(:disabled) {
  @apply transform scale-105; /* 悬停时放大 */
}

.native-button:disabled {
  @apply bg-gray-300 text-gray-500 cursor-not-allowed; /* 禁用状态 */
}

.native-button:active:not(:disabled) {
  @apply scale-95; /* 按下时缩小 */
}

/* 加载状态 */
.loading {
  @apply cursor-wait pointer-events-none; /* 显示加载时禁用点击 */
}