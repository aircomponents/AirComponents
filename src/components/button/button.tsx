import { Component, h, Host, Prop, State,Event,EventEmitter} from "@stencil/core";

@Component({
  tag: "air-button",
  styleUrl: "button.css",
  shadow: true,
})
export class AirButton {
  @Prop() size: "small" | "medium" | "large" = "medium";
  @Prop() variant:
    | "solid"
    | "outline"
    | "text"
    | "neo"
    | "default"="default";
  @Prop() color:
    | "default"
    | "primary"
    | "success"
    | "info"
    | "warning"
    | "danger"
    | "ghost"="default";
  @Prop() icon: string = "";
  @Prop() suffixIcon: string = "";
  @Prop() disabled: boolean = false;
  @Prop() loading: boolean = false; // 加载状态
  @Prop() selected: boolean = false; // 选中状态

  @State() hasSlotContent: boolean = false; // 插槽内容检测

  @Event() buttonClick: EventEmitter<{ event: MouseEvent; selected: boolean }>; // 定义事件

  private nativeElement!: HTMLButtonElement;
  // 检查按钮插槽内容
  private computeSlotHasContent() {
    const slot = this.nativeElement.shadowRoot?.querySelector("slot");
    if (slot) {
      const assignedNodes = slot.assignedNodes();
      this.hasSlotContent = assignedNodes.length > 0 || slot.textContent?.trim().length > 0;
    }
  }
  private handleClick = (event: MouseEvent) => {
    if (this.disabled || this.loading) {
      event.preventDefault();
      return;
    }

    // 触发外部注入的逻辑
    this.buttonClick.emit({ event, selected: this.selected });
  };

  render() {
    const ariaLabel = this.hasSlotContent ? null : "Button"; // 适配无内容时的 aria-label

    return (
      <Host>
        <button
          class={{
            "native-button": true,
            [`size-${this.size}`]: true,
            [`variant-${this.variant}`]: !!this.variant,
            [`color-${this.color}`]: true,
            "loading": this.loading, // 加载状态样式
            "disabled": this.disabled, // 禁用状态样式
            "selected": this.selected, // 选中状态样式
          }}
          aria-busy={this.loading ? "true" : null}
          aria-label={ariaLabel}
          title={ariaLabel}
          ref={(elm: HTMLButtonElement) => (this.nativeElement = elm)}
          disabled={this.disabled} // 禁用按钮
          onClick={this.handleClick}
        >
          {this.loading && <span class="native-button__loading-icon" /> }
          {!this.loading && this.icon && <span class="native-button__icon">{this.icon}</span>}
          <slot onSlotchange={() => this.computeSlotHasContent()} />
          {!this.loading && this.suffixIcon && <span class="native-button__suffix-icon">{this.suffixIcon}</span>}
        </button>
      </Host>
    );
  }
}