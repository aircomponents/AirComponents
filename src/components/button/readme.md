# air-button



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type                                                                                | Default     |
| ------------ | ------------- | ----------- | ----------------------------------------------------------------------------------- | ----------- |
| `color`      | `color`       |             | `"danger" \| "default" \| "ghost" \| "info" \| "primary" \| "success" \| "warning"` | `"default"` |
| `disabled`   | `disabled`    |             | `boolean`                                                                           | `false`     |
| `icon`       | `icon`        |             | `string`                                                                            | `""`        |
| `loading`    | `loading`     |             | `boolean`                                                                           | `false`     |
| `selected`   | `selected`    |             | `boolean`                                                                           | `false`     |
| `size`       | `size`        |             | `"large" \| "medium" \| "small"`                                                    | `"medium"`  |
| `suffixIcon` | `suffix-icon` |             | `string`                                                                            | `""`        |
| `variant`    | `variant`     |             | `"default" \| "neo" \| "outline" \| "solid" \| "text"`                              | `"default"` |


## Events

| Event         | Description | Type                                                     |
| ------------- | ----------- | -------------------------------------------------------- |
| `buttonClick` |             | `CustomEvent<{ event: MouseEvent; selected: boolean; }>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
