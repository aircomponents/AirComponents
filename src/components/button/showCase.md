<!-- 按钮颜色展示 -->
<air-button color="primary">主按钮</air-button>
<air-button color="success">成功按钮</air-button>
<air-button color="info">信息按钮</air-button>
<air-button color="warning">警告按钮</air-button>
<air-button color="danger">危险按钮</air-button>
<air-button color="ghost">幽灵按钮</air-button>

<!-- 按钮样式展示 -->
<air-button variant="solid">实心按钮</air-button>
<air-button variant="outline">描边按钮</air-button>
<air-button variant="text">文字按钮</air-button>
<air-button variant="neo">霓虹按钮</air-button>
<air-button variant="default">默认按钮</air-button>

<!-- 图标按钮展示 -->
<air-button icon="⭐">图标按钮</air-button>
<air-button suffix-icon="✓">后缀图标按钮</air-button>
<air-button icon="🔄" suffix-icon="✓">双图标按钮</air-button>

<!-- 加载中、禁用、选中状态展示 -->
<air-button loading>加载中</air-button>
<air-button disabled>禁用按钮</air-button>
<air-button disabled icon="❌">禁用图标按钮</air-button>
<air-button selected>选中按钮</air-button>
<air-button selected icon="✔️">选中图标按钮</air-button>

<!-- 不同大小按钮展示 -->
<air-button size="small">小按钮</air-button>
<air-button size="medium">中按钮</air-button>
<air-button size="large">大按钮</air-button>
