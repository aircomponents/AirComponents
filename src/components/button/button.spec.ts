import { newSpecPage } from "@stencil/core/testing";
import { AirButton } from "./button";

describe("air-button", () => {
  it("renders default air-button", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button></air-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("applies size prop correctly", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button size="small"></air-button>`,
    });
    const button = page.root.shadowRoot.querySelector("button");
    expect(button.classList.contains("size-small")).toBeTruthy();
  });

  it("applies variant prop correctly", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button variant="outline"></air-button>`,
    });
    const button = page.root.shadowRoot.querySelector("button");
    expect(button.classList.contains("variant-outline")).toBeTruthy();
  });

  it("applies color prop correctly", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button color="primary"></air-button>`,
    });
    const button = page.root.shadowRoot.querySelector("button");
    expect(button.classList.contains("color-primary")).toBeTruthy();
  });

  it("disables the button when disabled prop is true", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button disabled></air-button>`,
    });
    const button = page.root.shadowRoot.querySelector("button");
    expect(button.hasAttribute("disabled")).toBeTruthy();
    expect(button.classList.contains("disabled")).toBeTruthy();
  });

  it("shows loading icon when loading is true", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button loading></air-button>`,
    });
    const button = page.root.shadowRoot.querySelector("button");
    expect(button.getAttribute("aria-busy")).toBe("true");
    const loadingIcon = button.querySelector(".native-button__loading-icon");
    expect(loadingIcon).not.toBeNull();
  });

  it("renders icon and suffixIcon correctly", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button icon="✔" suffix-icon="❌"></air-button>`,
    });
    const icon = page.root.shadowRoot.querySelector(".native-button__icon");
    const suffixIcon = page.root.shadowRoot.querySelector(".native-button__suffix-icon");
    expect(icon.textContent).toBe("✔");
    expect(suffixIcon.textContent).toBe("❌");
  });

  it("handles slot content correctly", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button>Click Me</air-button>`,
    });
    expect(page.root.shadowRoot.querySelector("slot")).not.toBeNull();
    expect(page.root.textContent).toContain("Click Me");
  });

  it("updates aria-label when slot is empty", async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button></air-button>`,
    });
    const button = page.root.shadowRoot.querySelector("button");
    expect(button.getAttribute("aria-label")).toBe("Button");
  });

  it('emits buttonClick event on click', async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button>Click Me</air-button>`,
    });

    const button = page.root.shadowRoot.querySelector('button');
    const buttonClickSpy = jest.fn();

    // 监听自定义事件
    page.root.addEventListener('buttonClick', buttonClickSpy);

    // 模拟点击
    button.click();

    // 验证事件被触发
    expect(buttonClickSpy).toHaveBeenCalled();
  });

  it('does not emit buttonClick event when disabled', async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button disabled>Disabled Button</air-button>`,
    });

    const button = page.root.shadowRoot.querySelector('button');
    const buttonClickSpy = jest.fn();

    // 监听自定义事件
    page.root.addEventListener('buttonClick', buttonClickSpy);

    // 模拟点击
    button.click();

    // 验证事件未触发
    expect(buttonClickSpy).not.toHaveBeenCalled();
  });

  it('does not emit buttonClick event when loading', async () => {
    const page = await newSpecPage({
      components: [AirButton],
      html: `<air-button loading>Loading Button</air-button>`,
    });

    const button = page.root.shadowRoot.querySelector('button');
    const buttonClickSpy = jest.fn();

    // 监听自定义事件
    page.root.addEventListener('buttonClick', buttonClickSpy);

    // 模拟点击
    button.click();

    // 验证事件未触发
    expect(buttonClickSpy).not.toHaveBeenCalled();
  });
});

