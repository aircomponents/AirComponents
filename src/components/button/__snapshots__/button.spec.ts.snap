// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`air-button renders default air-button 1`] = `
<air-button>
  <template shadowrootmode="open">
    <button aria-label="Button" class="color-default native-button size-medium variant-default" title="Button">
      <slot></slot>
    </button>
  </template>
</air-button>
`;
